File: PRACTICA3/QytetetJava/src/modeloqytetet/Qytetet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

package modeloqytetet;

import java.util.ArrayList;
import java.util.Random;
import java.util.Scanner;

/**
 *
 * @author Jesus Manuel Garcia Palma
 * @correo xexugarcia95@gmail.com / jesusluisbcrespo@correo.ugr.es
 */
public class Qytetet {
    private ArrayList<Sorpresa> mazo = new ArrayList<>();
    private Tablero tablero;
    private static final Qytetet instance = new Qytetet();
    public int MAX_JUGADORES = 4;
    int NUM_SORPRESAS = 10;
    public int NUM_CASILLAS = 20;
    int PRECIO_LIBERTAD = 200;
    int SALDO_SALIDA = 1000;
    private Sorpresa cartaActual;
    private Jugador jugadorActual;
    private ArrayList<Jugador> jugadores = new ArrayList<>();
    private Dado dado;
    private EstadoJuego estado;
    private static final Scanner in= new Scanner(System.in);
    
    private Qytetet()
    {
       ArrayList<String> nombres = new ArrayList();
       nombres.add("Jugador 1");
       nombres.add("Jugador 2");
       inicializarJuego(nombres);
    }
    
    public static Qytetet getInstance()
    {
        return instance;
    }
    
   
    private void inicializarTablero()
    {
        tablero = new Tablero();
    }
    
    void actuarSiEnCasillaEdificable()
    {
        boolean deboPagar = jugadorActual.deboPagarAlquiler();
        if(deboPagar)
        {
            jugadorActual.pagarAlquiler();
            if(jugadorActual.getSaldo()<0)
            {
                setEstado(estado.ALGUNJUGADORENBANCARROTA);
            }
        }
        Casilla casilla = obtenerCasillaJugadorActual();
        boolean tengoPropietario = casilla.tengoPropietario();
        
        if(estado==EstadoJuego.ALGUNJUGADORENBANCARROTA)
        {
            if(tengoPropietario)
            {
                setEstado(EstadoJuego.JA_PUEDEGESTIONAR);
            }else
            {
                setEstado(EstadoJuego.JA_PUEDECOMPRAROGESTIONAR);
            }
        }
    }
    
    void actuarSiEnCasillaNoEdificable()
    {
        setEstado(EstadoJuego.JA_PUEDEGESTIONAR);
        Casilla casillaActual = jugadorActual.getCasillaActual();
        if(casillaActual.getTipo()==TipoCasilla.IMPUESTO)
        {
            jugadorActual.pagarImpuesto();
        }else if(casillaActual.getTipo()==TipoCasilla.JUEZ)
        {
            encarcelarJugador();
        }else if(casillaActual.getTipo()==TipoCasilla.SORPRESA)
        {
            mazo.remove(cartaActual);
            setEstado(EstadoJuego.JA_CONSORPRESA);
        }
    }
    
    public void aplicarSorpresa()
    {
        setEstado(EstadoJuego.JA_PUEDEGESTIONAR);
        if(cartaActual.getTipo()==TipoSorpresa.SALIRCARCEL)
        {
            jugadorActual.setCartaLibertad(cartaActual);
        }else
        {
            mazo.add(cartaActual); //Incluir al final
            if(cartaActual.getTipo()==TipoSorpresa.PAGARCOBRAR)
            {
                jugadorActual.modificarSaldo(cartaActual.getValor());
                if(jugadorActual.getSaldo()<0)
                {
                    setEstado(EstadoJuego.ALGUNJUGADORENBANCARROTA);
                }
            }else if(cartaActual.getTipo()==TipoSorpresa.IRACASILLA)
                {
                    int valor = cartaActual.getValor();
                    boolean casillaCarcel = tablero.esCasillaCarcel(valor);
                    if(casillaCarcel)
                    {
                        encarcelarJugador();
                    }else
                    {
                        mover(valor);
                    }
                }else if(cartaActual.getTipo()==TipoSorpresa.PORCASAHOTEL)
                {
                    int cantidad = cartaActual.getValor();
                    int numeroTotal = jugadorActual.cuantasCasasHotelesTengo();
                    jugadorActual.modificarSaldo(cantidad*numeroTotal);
                    
                    if(jugadorActual.getSaldo()<0)
                    {
                        setEstado(EstadoJuego.ALGUNJUGADORENBANCARROTA);
                    }
                }else if(cartaActual.getTipo()==TipoSorpresa.PORJUGADOR)
                {
                    for(int i=0;i<MAX_JUGADORES-1;i++)
                    {
                        Jugador jugador = jugadores.get(i);
                        if(jugador!=jugadorActual)
                        {
                            jugadores.get(i).modificarSaldo(cartaActual.getValor());
                        }
                        if(jugador.getSaldo()<0)
                        {
                            setEstado(EstadoJuego.ALGUNJUGADORENBANCARROTA);
                        }
                        
                        jugadorActual.modificarSaldo(-cartaActual.getValor());
                        if(jugadorActual.getSaldo()<0)
                        {
                            setEstado(EstadoJuego.ALGUNJUGADORENBANCARROTA);
                        }
                    }
                }
            
        }
    }
    
    public boolean cancelarHipoteca(int numeroCasilla)
    {
        Casilla casilla = tablero.getCasillas().get(numeroCasilla);
        TituloPropiedad titulo = casilla.getTitulo();
        boolean cancelar = jugadorActual.cancelarHipoteca(titulo);
        setEstado(EstadoJuego.JA_PUEDEGESTIONAR);
        return cancelar;
    }
    
    public boolean comprarTituloPropiedad() 
    {
        boolean comprado = jugadorActual.comprarTituloPropiedad();
        if(comprado) setEstado(EstadoJuego.JA_PUEDEGESTIONAR);
        
        return comprado;
    }
    
    public boolean edificarCasa(int numeroCasilla)
    {
       Casilla casilla = tablero.obtenerCasillaNumero(numeroCasilla);
       TituloPropiedad titulo = casilla.getTitulo();
       boolean edificada = jugadorActual.edificarCasa(titulo);
       return edificada;
    }
    
    public boolean edificarHotel(int numeroCasilla)
    {
        Casilla casilla = tablero.obtenerCasillaNumero(numeroCasilla);
       TituloPropiedad titulo = casilla.getTitulo();
       boolean edificada = jugadorActual.edificarHotel(titulo);
       return edificada;
    }
    
    private void encarcelarJugador()
    {
        if(!jugadorActual.tengoCartaLibertad())
        {
            Casilla casillaCarcel = tablero.getCarcel();
            jugadorActual.irACarcel(casillaCarcel);
            setEstado(EstadoJuego.JA_ENCARCELADO);
        }else
        {
            Sorpresa carta = jugadorActual.devolverCartaLibertad();
            mazo.add(carta); //incluir al final
            setEstado(EstadoJuego.JA_PUEDEGESTIONAR);
        }
    }

    public int getValorDado()
    {
        return dado.getValor();
    }
    
    public void hipotecarPropiedad(int numeroCasilla)
    {
        Casilla casilla = tablero.obtenerCasillaNumero(numeroCasilla);
        TituloPropiedad titulo = casilla.getTitulo();
        jugadorActual.hipotecarPropiedad(titulo);
        setEstado(EstadoJuego.JA_PUEDEGESTIONAR);
    }
    
    public void inicializarJuego(ArrayList<String> nombres)
    {
        inicializarJugadores(nombres);
        inicializarCartasSorpresa();
//        salidaJugadores();
        
    }
    
    private void inicializarJugadores(ArrayList<String> nombres)
    {
        for(int i=0;i<nombres.size();i++)
        { 
            Jugador j = new Jugador(nombres.get(i));
            jugadores.add(j);
        }
    }
    
    public boolean intentarSalirCarcel(MetodoSalirCarcel metodo)
    {
        if(metodo==MetodoSalirCarcel.TIRANDODADO)
        {
            int resultado = tirarDado();
            if(resultado>=5) jugadorActual.setEncarcelado(false);            
        }else if(metodo==MetodoSalirCarcel.PAGANDOLIBERTAD)
        {
            jugadorActual.pagarLibertad(PRECIO_LIBERTAD);
            
        }
        
        boolean libre = jugadorActual.getEncarcelado();
        
        if(libre)
        {
            setEstado(EstadoJuego.JA_ENCARCELADO);
        }else
        {
            setEstado(EstadoJuego.JA_PREPARADO);
        }
        
        return libre;
    }
    
    public void jugar()
    {
        int i =tirarDado();
        Casilla casillaActual= tablero.obtenerCasillaFinal(jugadorActual.getCasillaActual(), i);
        mover(casillaActual.getNumeroCasilla());
    }
    
    void mover(int numCasillaDestino)
    {
        Casilla casillaInicial = jugadorActual.getCasillaActual();
        Casilla casillaFinal = tablero.obtenerCasillaNumero(numCasillaDestino);
        jugadorActual.setCasillaActual(casillaFinal);
        
        if(numCasillaDestino<casillaInicial.getNumeroCasilla())
        {
            jugadorActual.modificarSaldo(SALDO_SALIDA);
        }
        
        if(casillaFinal.soyEdificable())
        {
            actuarSiEnCasillaEdificable();
        }else
        {
            actuarSiEnCasillaNoEdificable();
        }
    }
    
    public Casilla obtenerCasillaJugadorActual()
    {
        return jugadorActual.getCasillaActual();
    }
    
    public ArrayList<Casilla> obtenerCasillasTablero()
    {
        return tablero.getCasillas();
    }
    
    public ArrayList<Integer> obtenerPropiedadesJugador()
    {
        ArrayList<Integer> numPropiedades = new ArrayList<>();
        for(int i=0;i<jugadorActual.getPropiedades().size();i++)
        {
            boolean encontrado = false;
            for(int j=0;j<tablero.getCasillas().size() && !encontrado;j++)
            {
                if(jugadorActual.getPropiedades().get(i)==tablero.getCasillas().get(j).getTitulo())
                {
                    numPropiedades.add(tablero.getCasillas().get(j).getNumeroCasilla());
                    encontrado = true;
                }
            }
                
        }
        return numPropiedades;
    }
    
    public ArrayList<Integer> obtenerPropiedadesJugadorSegunEstadoHipoteca(boolean estadoHipoteca)
    {
        ArrayList<Integer> numPropiedades = new ArrayList<>();
        for(int i=0;i<jugadorActual.getPropiedades().size();i++)
        {
            boolean encontrado = false;
            for(int j=0;j<tablero.getCasillas().size() && !encontrado;j++)
            {
                if(jugadorActual.getPropiedades().get(i)==tablero.getCasillas().get(j).getTitulo() 
                        && jugadorActual.getPropiedades().get(j).isHipotecada() == estadoHipoteca)
                {
                    numPropiedades.add(tablero.getCasillas().get(j).getNumeroCasilla());
                    encontrado = true;
                }
            }
                
        }
        return numPropiedades;
    }
    
    public void obtenerRanking()
    {
        ArrayList<Jugador> posiciones = jugadores;
        for(int i=0;i<posiciones.size();i++)
        {
            for(int j=i+1;j<posiciones.size()-1;j++)
            {
                if(posiciones.get(i).compareTo(posiciones.get(j))<=0)
                {
                    Jugador aux = posiciones.get(j);
                    posiciones.set(j,posiciones.get(i));
                    posiciones.set(i,aux);
                }
            }
        }
        
        for(int i=0;i<posiciones.size();i++)
        {
            System.out.println("Posicion " + i+1 + ": " + posiciones.get(i));
        }
    }
    
    public int obtenerSaldoJugadorActual()
    {
        return jugadorActual.getSaldo();
    }
    
    private void salidaJugadores()
    {
        for(int i=0;i<jugadores.size();i++) jugadores.get(i).setCasillaActual(tablero.obtenerCasillaNumero(0));
        Random r = new Random();
        int numero = r.nextInt(1);
        jugadorActual = jugadores.get(numero);
        estado = EstadoJuego.JA_PREPARADO;
    }
    
    public void siguienteJugador()
    {
        int mod = 0;
        boolean encontrado = false;
        for(int i=0;i<jugadores.size() && !encontrado;i++)
        {
            if(jugadores.get(i)==jugadorActual) mod = (i+1)%jugadores.size(); encontrado = true;
        }
        
        jugadorActual = jugadores.get(mod);
        if(jugadorActual.isEncarcelado()) estado = EstadoJuego.JA_ENCARCELADOCONOPCIONDELIBERTAD;
        else estado = EstadoJuego.JA_PREPARADO;
    }
    
    public Jugador siguiente()
    {
        int mod = 0;
        boolean encontrado = false;
        for(int i=0;i<jugadores.size() && !encontrado;i++)
        {
            if(jugadores.get(i)==jugadorActual) mod = (i+1)%jugadores.size(); encontrado = true;
        }
        
        return jugadores.get(mod);
    }
    
    int tirarDado()
    {
        return dado.tirar();
    }
    
    public boolean venderPropiedad(int numeroCasilla)
    {
        Casilla casilla = tablero.obtenerCasillaNumero(numeroCasilla);
        boolean puedeVender = jugadorActual.venderPropiedad(casilla);
        if(puedeVender) setEstado(EstadoJuego.JA_PUEDEGESTIONAR);
        return puedeVender;
        
    }
    
    
    
    
    void inicializarCartasSorpresa()
    {
        inicializarTablero();
        mazo.add(new Sorpresa("El banco tira la casa por la ventana. Recibe 500",TipoSorpresa.PAGARCOBRAR,500));
        mazo.add(new Sorpresa("Multa por bailar la conga. Paga 500",TipoSorpresa.PAGARCOBRAR,-500));
        mazo.add(new Sorpresa("Corre como el viento perdigón!. Ve a la casilla de salida.",TipoSorpresa.IRACASILLA,0));
        mazo.add(new Sorpresa("Te posee el espíritu de M.J y bailas un Moonwalk. Retrocede a la casilla 5.",TipoSorpresa.IRACASILLA,5));
        mazo.add(new Sorpresa("Te ha pillado la pasma. Vas a la carcel",TipoSorpresa.IRACASILLA,tablero.getCarcel().getNumeroCasilla()));
        mazo.add(new Sorpresa("El banco ha errado en la hipoteca. Recibe 50 por cada casa/hotel.",TipoSorpresa.PORCASAHOTEL,50));
        mazo.add(new Sorpresa("Te se ha ido la olla hipotecando inmuebles. Paga 50 por cada casa/hotel",TipoSorpresa.PORCASAHOTEL,-50));
        mazo.add(new Sorpresa("Es tu cumpleaños. Recibe 20 de cada jugador.",TipoSorpresa.PORJUGADOR,20));
        mazo.add(new Sorpresa("Has perdido una apuesta. Paga 20 a cada jugador.",TipoSorpresa.PORJUGADOR,-20));
        mazo.add(new Sorpresa("Has quedado libre de la cárcel.",TipoSorpresa.SALIRCARCEL,0));
    }
    
    
    public ArrayList<String> getNombreJugadores()
    {
        
        ArrayList<String> array = new ArrayList<>();
        for(Jugador j: jugadores)
        {
            System.out.println(j.getNombre());
            String s = in.nextLine();
            System.out.println(s);
            array.add(s);
        }
        
        return array;
    }

    public Tablero getTablero() {
        return tablero;
    }
    
    public Sorpresa getCartaActual() {
        return cartaActual;
    }

    private Jugador getJugadorActual() {
        return jugadorActual;
    }
    
    public String getNombreJugadorActual()
    {
        return jugadorActual.getNombre();
    }

    public ArrayList<Jugador> getJugadores() {
        return jugadores;
    }

    Dado getDado() {
        return dado;
    }
    
     ArrayList<Sorpresa> getMazo() {
        return mazo;
    }

    private void setCartaActual(Sorpresa cartaActual) {
        this.cartaActual = cartaActual;
    }

    public EstadoJuego getEstado() {
        return estado;
    }

    public void setEstado(EstadoJuego estado) {
        this.estado = estado;
    }
    
    

    @Override
    public String toString() {
        return "Qytetet{" + "mazo=" + mazo + ", tablero=" + tablero + ", MAX_JUGADORES=" + MAX_JUGADORES + ", NUM_SORPRESAS=" + NUM_SORPRESAS + ", NUM_CASILLAS=" + NUM_CASILLAS + ", PRECIO_LIBERTAD=" + PRECIO_LIBERTAD + ", SALDO_SALIDA=" + SALDO_SALIDA + ", cartaActual=" + cartaActual + ", jugadorActual=" + jugadorActual + ", jugadores=" + jugadores + ", dado=" + dado + '}';
    }


     
     
    
}

File: PRACTICA3/QytetetRuby/lib/qytetet.rb
#encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative 'sorpresa.rb'
require_relative 'tablero.rb'
require_relative 'jugador.rb'
require_relative 'estado_juego.rb'


module ModeloQytetet
  require 'singleton'
  class Qytetet
    include Singleton


    attr_reader :mazo, :tablero
    #@@instance = Qytetet.instance
    @mazo = nil
    @tablero = nil
    @jugadores = nil
    @@max_jugadores = 4
    @@num_sorpresas = 10
    @@num_casillas = 20
    @@precio_libertad = 200
    @@saldo_salida = 1000
    @carta_actual
    @jugador_actual
    @estado
    @dado

    def inicializar_cartas_sorpresa
        @mazo = Array.new
        inicializar_tablero
        @mazo << Sorpresa.new("El banco tira la casa por la ventana. Recibe 500",TipoSorpresa::PAGARCOBRAR,500)
        @mazo << Sorpresa.new("Multa por bailar la conga. Paga 500",TipoSorpresa::PAGARCOBRAR,-500)
        @mazo << Sorpresa.new("Corre como el viento perdigón!. Ve a la casilla de salida.",TipoSorpresa::IRACASILLA,0)
        @mazo << Sorpresa.new("Te posee el espíritu de M.J y bailas un Moonwalk. Retrocede a la casilla 5.",TipoSorpresa::IRACASILLA,5)
        @mazo << Sorpresa.new("Vas pegando brincos de alegría. Ve a la casilla 15.",TipoSorpresa::IRACASILLA,15)
        @mazo << Sorpresa.new("El banco ha errado en la hipoteca. Recibe 50 por cada casa/hotel.",TipoSorpresa::PORCASAHOTEL,50)
        @mazo << Sorpresa.new("Te se ha ido la olla hipotecando inmuebles. Paga 50 por cada casa/hotel",TipoSorpresa::PORCASAHOTEL,-50)
        @mazo << Sorpresa.new("Es tu cumpleaños. Recibe 20 de cada jugador.",TipoSorpresa::PORJUGADOR,20)
        @mazo << Sorpresa.new("Has perdido una apuesta. Paga 20 a cada jugador.",TipoSorpresa::PORJUGADOR,-20)
        @mazo << Sorpresa.new("Has quedado libre de la cárcel.",TipoSorpresa::SALIRCARCEL,0)
    end

    def actuar_si_en_casilla_edificable
      debo_pagar = jugadorActual.debo_pagar_alquiler

        if(debo_pagar)
        
            jugadorActual.pagar_alquiler
            if(jugadorActual.saldo<0)
                @estado = EstadoJuego::ALGUNJUGADORENBANCARROTA              
            end
        end
        casilla = obtener_casilla_jugador_actual
        tengo_propietario = casilla.tengo_propietario
        
        if(@estado==EstadoJuego::ALGUNJUGADORENBANCARROTA)
        
            if(tengo_propietario)
                @estado = EstadoJuego::JA_PUEDEGESTIONAR
            else
                @estado = EstadoJuego::JA_PUEDECOMPRARGESTIONAR
            end
        end    
    end

    def actuar_si_en_casilla_no_edificable
      @estado=EstadoJuego::JA_PUEDEGESTIONAR
        @casilla_actual = @jugador_actual.casilla_actual
        if(@casilla_actual.tipo==TipoCasilla::IMPUESTO)
        
            jugadorActual.pagarImpuesto();
        elsif(@casilla_actual.tipo==TipoCasilla::JUEZ)
        
            encarcelar_jugador
        elsif(@casilla_actual.tipo==TipoCasilla::SORPRESA)
        
            @mazo.remove(@carta_actual);
            @estado = EstadoJuego::JA_CONSORPRESA
        end
    end


    def aplicar_sorpresa
     @estado=EstadoJuego::JA_PUEDEGESTIONAR
        if(@carta_actual.tipo==TipoSorpresa::SALIRCARCEL)
        
            @jugador_actual.carta_libertad = @carta_actual
        else
        
            @mazo << @carta_actual
            if(@carta_actual.tipo==TipoSorpresa::PAGARCOBRAR)
            
                jugador_actual.modificar_saldo(@carta_actual.valor)
                if(jugador_actual.saldo<0)
                  
                    @estado = EstadoJuego::ALGUNJUGADORENBANCARROTA
                end
            elsif(@carta_actual.tipo==TipoSorpresa::IRACASILLA)
                
                    valor = @carta_actual.valor
                    casilla_carcel = @tablero.es_casilla_carcel(valor)
                    if(casilla_carcel)
                    
                        encarcelar_jugador
                    else
                    
                        mover(valor)
                    end
            elsif(@carta_actual.tipo==TipoSorpresa::PORCASAHOTEL)
                
                    cantidad = @carta_actual.valor
                    numero_total = @jugador_actual.cuantas_casas_hoteles_tengo
                    @jugador_actual.modificar_saldo(cantidad*numero_total);
                    
                    if(@jugador_actual.saldo<0)
                    
                        @estado= EstadoJuego::ALGUNJUGADORENBANCARROTA
                    end
            elsif(@carta_actual.tipo==TipoSorpresa::PORJUGADOR)
                
                    for i in 0..@max_jugadores
                    
                        jugador = @jugadores[i]
                        if(jugador!=@jugador_actual)
                        
                            jugadores[i].modificar_saldo(@carta_actual.valor)
                        end
                        if(jugador.saldo<0)
                        
                            @estado = EstadoJuego::ALGUNJUGADORENBANCARROTA
                        end
                        
                        @jugador_actual.modificar_saldo(-@carta_actual.valor)
                        if(@jugador_actual.saldo<0)
                        
                            @estado= EstadoJuego::ALGUNJUGADORENBANCARROTA
                        end
                    end 
            end
        end   
    end


    def cancelar_hipoteca(numero_casilla)

        casilla = @tablero.casillas[numero_casilla]
        titulo = casilla.titulo
        cancelar = @jugador_actual.cancelar_hipoteca(titulo)
        @estado=EstadoJuego::JA_PUEDEGESTIONAR
        return cancelar
    end

    def comprar_titulo_propiedad

        comprado = @jugador_actual.comprar_titulo_propiedad
        if(comprado) 
          @estado = EstadoJuego::JA_PUEDEGESTIONAR
        end
        
        return comprado
    end

    def edificar_casa(numero_casilla)

        casilla = @tablero.obtener_casilla_numero(numero_casilla)
        titulo = casilla.titulo
        edificada = @jugador_actual.edificar_casa(titulo)
       return edificada
    end

    def edificar_hotel(numero_casilla)

       casilla = @tablero.obtener_casilla_numero(numero_casilla);
       titulo = casilla.titulo
       edificada = @jugador_actual.edificar_hotel(titulo)
       return edificada
    end



    def get_valor_dado

        return @dado.valor
    end

    def hipotecar_propiedad(numero_casilla)
        casilla = @tablero.obtener_casilla_numero(numero_casilla);
        titulo = casilla.titulo
        @jugador_actual.hipotecar_propiedad(titulo)
        @estado= EstadoJuego::JA_PUEDEGESTIONAR
    end


    def inicializar_juego(nombres)
        inicializar_cartas_sorpresa
        inicializar_jugadores(nombres)
        
    end



    def intentar_salir_carcel(metodo)
    
        if(metodo == MetodoSalirCarcel::TIRANDODADO)
        
            resultado = tirar_dado
            if(resultado>=5) 
              @jugador_actual.encarcelado = false
            end
        elsif(metodo == MetodoSalirCarcel::PAGANDOLIBERTAD)
        
            @jugador_actual.pagar_libertad(@@precio_libertad)          
        end
        
        libre = @jugador_actual.encarcelado
        
        if(libre)
        
            @estado = EstadoJuego.JA_ENCARCELADO
        else
        
            @estadi = EstadoJuego.JA_PREPARADO
        end
        
        return libre
    end

    def jugar
        i =  tirar_dado
        casilla_actual = @tablero.obtener_casilla_final(@jugadorActual.casilla_actual,i)
        mover(casilla_actual.numero_casilla)

    end

    def mover(num_casilla_destino)
        casilla_inicial = @jugador_actual.casilla_actual
        casilla_final = @tablero.obtener_casilla_numero(num_casilla_destino)
        @jugador_actual.casilla_actual = casilla_final
        
        if(num_casilla_destino<casilla_inicial.numero_casilla)
        
            @jugador_actual.modificar_saldo(@@saldo_salida)
        end
        
        if(casilla_final.soy_edificable)
        
            actuar_si_en_casilla_edificable
        else
        
            actuar_si_en_casilla_no_edificable
        end

    end

    def obtener_casilla_jugador_actual

        return @jugador_actual.casilla_actual
    end

    def obtener_casillas_tablero

        return @tablero.casillas
    end

    def obtener_propiedades_jugador

        num_propiedades = Array.new
        for i in 0..@jugador_actual.propiedades.size
        
            encontrado = false
            for j in 0..@tablero.casillas.size && !encontrado
            
                if(@jugador_actual.propiedades[i]==@tablero.casillas[i].titulo)
                
                    num_propiedades << @tablero.casillas[j].numero_casilla
                    encontrado = true;
                end
            end
                
        end
        return numPropiedades
    end

    def obtener_propiedades_jugador_segun_estado_hipoteca(estado_hipoteca)

        num_propiedades = Array.new
        for i in 0..@jugador_actual.propiedades.size
        
            encontrado = false
            for j in 0..@tablero.casillas.size && !encontrado
            
                if(@jugador_actual.propiedades[i]==@tablero.casillas[j].titulo && @jugador_actual.propiedades[j].is_hipotecada == estado_hipoteca)
                
                    num_propiedades << @tablero.casillas[j].numero_casilla
                    encontrado = true
                end
            end
                
        end
        return num_propiedades
    end

    def obtener_ranking
      posiciones = @jugadores
        for i in 0..posiciones.size
        
            for j in i+1..@posiciones.size-1
            
                if(posiciones[i].<=>(posiciones[j])<=0)
                
                    aux = posiciones[j]
                    posiciones[j] = posiciones[i]
                    posiciones[i] = aux
                end
            end
        end
        
        for i in 0..@posiciones.size
            puts ("Posicion " + i+1 + ": " + @posiciones[i])
        end

    end

    def  obtener_saldo_jugador_actual

        return @jugador_actual.saldo
    end



    def siguiente_jugador
        mod = 0
        encontrado = false
        for i in 0..@jugadores.size && !encontrado
       
            if(jugadores[i]==@jugador_actual) 
              mod = (i+1)%@jugadores.size 
              encontrado = true;
            end
        end
        
        @jugador_actual = @jugadores[mod]
        if(@jugador_actual.is_encarcelado)
          @stado = EstadoJuego::JA_ENCARCELADOCONOPCIONDELIBERTAD
        else 
          @estado = EstadoJuego::JA_PREPARADO
        end

    end
    
    def siguiente
        mod = 0;
        encontrado = false
        for i in 0..@jugadores.size && !encontrado
        
            if(@jugadores[i]==@jugador_actual) 
              mod = (i+1)%@jugadores.size 
              encontrado = true
            end
        end
        return @jugadores[mod]
    end

    def tirar_dado
    
        @dado.tirar
    end

    def vender_propiedad(numero_casilla)

        casilla = @tablero.obtener_casilla_numero(numero_casilla)
        puede_vender = @jugador_actual.vender_propiedad(casilla)
        if(puede_vender) 
          @estado = EstadoJuego::JA_PUEDEGESTIONAR
        end
        return puede_vender
    end
    
    def nombre_jugadores
      array = Array.new
        jugadores.each do |j|
        
            puts(j.nombre)
            array << j.nombre
            
        end
        
        return array
    end
    
    def initialize
      @jugadores = Array.new
      @estado = EstadoJuego.new
      nombres = Array.new
      nombres<<"Jugador 1"
      nombres<<"Jugador 2"
      inicializar_juego(nombres)

    end
    
    def to_s
      @jugadores.each do |ju|
        puts ju
      end
      
      puts "Tablero: #{@tablero} \n"
      @mazo.each do |m|
        puts m
      end
    end

    private

    

    def inicializar_tablero
      @tablero = Tablero.new
    end

    def encarcelar_jugador
      if(!@jugador_actual.tengo_carta_libertad)
        
            casilla_carcel = @tablero.carcel
            @jugador_actual.ir_a_carcel(casilla_carcel)
            @estado = EstadoJuego.JA_ENCARCELADO
       else
        
            carta = @jugador_actual.devolver_carta_libertad
            @mazo << carta 
            @estado= EstadoJuego.JA_PUEDEGESTIONAR
      end

    end

    def inicializar_jugadores(nombres)

        nombres.each do |n|

            j = Jugador.new(n)
            @jugadores << j
            
        end

    end

    def salida_jugadores
      for i in 0..@jugadores.size 
        @jugadores[i].casilla_actual = @tablero.obtener_casilla_numero(0)
      end
        r = Random.new
        numero = r.rand(0...1)
        @jugador_actual = @jugadores[numero]
        @estado = EstadoJuego::JA_PREPARADO
    end
    

  end
end
